### PR TITLE
Irsa 2887 uint64

### PR DESCRIPTION
**Pull request by judith-ipac**
Hi, Serge.  The first of these commits is just formatting with clang.

### CODE REVIEW

File: src/Table.hxx [hunk 1]
```diff
@@ -108,6 +108,62 @@ public:
...
 
+    //====================================================================
+
+    inline std::map<size_t, Data_Type> determine_datatypes_for_writing(
+            const Format &format) const {
+        std::map<size_t, Data_Type> alt_datatype_lookup;
+        bool adjust_uint64 = ((format.enum_format == Format::Enums::VOTABLE) ||
+                              (format.enum_format == Format::Enums::IPAC_TABLE) ||
+                              (format.enum_format == Format::Enums::FITS));
+        bool adjust_all_unsigned = (format.enum_format == Format::Enums::VOTABLE);
```
Comment (smonkewitz) [inline src/Table.hxx]: So, I just want to point out that the FITS standard does in fact support unsigned integers in a roundabout way. See section 5.2.5 of [the standard](https://fits.gsfc.nasa.gov/standard40/fits_standard40aa-le.pdf) - the gist is that you set the (binary table) column type to be a signed m-bit integer (where m = 16, 32, or 64), and then specify an offset by setting TSCALn  = 1 and TZEROn = 2^(m - 1). You have to subtract 2^(m - 1) from unsigned values before storage to get appropriate signed values.

In other words, gymnastics are required for all unsigned integer types in FITS. But! As far as I can tell, CFITSIO handles the pain for you. Or at least the version I am looking at (3.45) `#defines` TUINT, TUSHORT, TULONGLONG, so I don't think you need to do any type conversions at a higher level.

File: src/Table.hxx [hunk 2]
```diff
@@ -108,6 +108,62 @@ public:
...
+    }
+
+    inline Data_Type get_datatype_for_writing(
+            const std::map<size_t, Data_Type> &datatypes_for_writing, size_t col) const {
+        auto iter = datatypes_for_writing.find(col);
+        if (iter == datatypes_for_writing.end()) {
+            return columns[col].type;
+        }
+        return iter->second;
+    }
```
Comment (smonkewitz) [inline src/Table.hxx]: Instead of a map from column index to datatypes, I think it would probably be simpler and more performant to use a vector of output datatypes.

File: src/Table/read_property_tree_as_votable/read_resource/read_table/read_data/read_tabledata/count_elements.cxx [hunk 3]
```diff
@@ -1,7 +1,6 @@
-#include <vector>
-
 #include <boost/algorithm/string.hpp>
 #include <boost/algorithm/string/trim_all.hpp>
+#include <vector>
```
Comment (smonkewitz) [inline src/Table/read_property_tree_as_votable/read_resource/read_table/read_data/read_tabledata/count_elements.cxx]: Standard library includes should precede boost includes.

File: src/Table/read_property_tree_as_votable/read_resource/read_table/read_data/read_tabledata/read_tabledata.cxx [hunk 4]
```diff
@@ -81,11 +81,17 @@ void tablator::Table::read_tabledata(const boost::property_tree::ptree &tabledat
...
+                                             ". Error message: " + error.what());
                 } catch (std::exception &error) {
                     throw std::runtime_error(
-                            "Invalid " + to_string(fields[column].type) + " in row " +
-                            std::to_string(current_row + 1) + ", field " +
-                            std::to_string(column) + ".  Found '" + element + "'");
+                            "Invalid " + to_string(fields[column].type) + "value " +
+                            element + " in row " + std::to_string(current_row + 1) +
+                            ", field " + std::to_string(column) +
+                            ". Error message: " + error.what());
```
Comment (smonkewitz) [inline src/Table/read_property_tree_as_votable/read_resource/read_table/read_data/read_tabledata/read_tabledata.cxx]: Did you mean "Failed to read value" instead of "Failed to write value"? Also, the responses to these 2 exception types are nearly identical. Can they be consolidated into a single catch?

File: src/Table.hxx [hunk 5]
```diff
@@ -166,13 +224,26 @@ public:
...
+
+    void write_tabledata(std::ostream &os, const Format::Enums &output_format) const {
+         write_tabledata(os, output_format, std::map<size_t, Data_Type>());
+    }
+
     void write_html(std::ostream &os) const;
     boost::property_tree::ptree generate_property_tree(
-            const std::string &tabledata_string) const;
+            const std::string &tabledata_string,
+            const std::map<size_t, Data_Type> &alternative_datatypes) const;
```
Comment (smonkewitz) [inline src/Table.hxx]: I'm probably missing something, but isn't letting users pass in `alternative_datatypes` in the various write methods (versus computing appropriate output types internally, in a non-publicly accessible way) a foot-gun?

File: src/Table.hxx (hunk 1 above)
Comment (smonkewitz) [reply src/Table.hxx]: Ugh. The version of CFITSIO in Debian Jessie doesn't have TULONGLONG; upgrading to Buster will fix that. So I guess we should keep the special handling for FITS, but add comments indicating the FITS workaround should be removed as soon as we have a sufficiently recent version of CFITSIO to play with.

File: .clang-format [hunk 6]
```diff
@@ -0,0 +1,10 @@
+---
+Language: Cpp
+BasedOnStyle: Google
+ColumnLimit: 88
+IndentWidth: 4
+AccessModifierOffset: -4
+SortIncludes: true
+ConstructorInitializerIndentWidth: 8
+ContinuationIndentWidth: 8
+...
```
[the remark this answers is not in the document]
Comment (judith-ipac) [reply .clang-format]: I tried that and it had no effect.  Will investigate.

File: src/Table.hxx (hunk 1 above)
Comment (judith-ipac) [reply src/Table.hxx]: Will do.

File: src/Table.hxx (hunk 2 above)
Comment (judith-ipac) [reply src/Table.hxx]: Okay.  For performance, I'm more worried about iterating through great big tables.

File: src/Table.hxx (hunk 5 above)
Comment (smonkewitz) [reply src/Table.hxx]: Good point - you convinced me we should leave it as is, and come back to it in a more principled way later.

File: test/recursive_param.xml [hunk 7]
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="utf-8"?>
-<VOTABLE version="1.3" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://www.ivoa.net/xml/VOTable/v1.3" xmlns:stc="http://www.ivoa.net/xml/STC/v1.30">
+<VOTABLE version="1.3" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://www.ivoa.net/xml/VOTable/v1.3" xmlns:stc="http://www.ivoa.net/xml/STC/v1.30" xsi:schemaLocation="http://www.ivoa.net/xml/VOTable/v1.3 http://www.ivoa.net/xml/VOTable/v1.3 http://www.ivoa.net/xml/STC/v1.30 http://www.ivoa.net/xml/STC/v1.30">
```
Comment (smonkewitz) [inline test/recursive_param.xml]: Thanks for the fix!

File: src/Table/write_ipac_table/write_ipac_table.cxx [hunk 8]
```diff
@@ -88,40 +99,37 @@ void tablator::Table::write_ipac_table(std::ostream &os) const {
     }
     os << "\n";
 
+    // Write column value
     std::stringstream ss;
     for (size_t row_offset = 0; row_offset < data.size(); row_offset += row_size()) {
         /// Skip the null bitfield flag
         for (size_t i = 1; i < columns.size(); ++i) {
-            for (size_t element = 0; element < columns[i].array_size; ++element) {
+            auto column = &(columns[i]);
```
[the remark this answers is not in the document]
Comment (judith-ipac) [reply src/Table/write_ipac_table/write_ipac_table.cxx]: Sure. I used a pointer so as not to have to change code below.

**Comment by judith-ipac** [comment]
Re "I'm still a little worried about table serialization methods taking very general looking maps of alternate datatypes to write...": Okay, I will try to think of a better way to manage the alternative datatypes.
Please check out my comments here https://jira.ipac.caltech.edu/browse/IRSA-2887 if you haven't recently.

**Comment by smonkewitz** [comment]
Looking at [to_xml_string.hxx](https://github.com/Caltech-IPAC/tablator/blob/IRSA-2887_UINT64/src/to_xml_string.hxx), I still see invalid datatypes ("ushort", "uint", "ulong"). If those were replaced ("ushort" -> "int", "uint" -> "long", "ulong" -> "long") couldn't you allow only uint64_t -> string remappings on output? (I.e. no need to handle the `adjust_all_unsigned` case in `Table.hxx`). At a glance I am not sure if that would simplify the code significantly though.

Alternatively,  the idea of just throwing an exception in response to a uint64_t values that will not fit in an int64_t is growing on me.

**Comment by judith-ipac** [comment]
I didn't want to mess with to_xml_string() because it is called for JSON and JSON5 as well as for VOTABLE.
Hmm, well, I'll keep going for the time being, but keep me posted about the extent that that idea is growing on you.

File: src/Data_Type_Adjuster.cxx [hunk 9]
```diff
@@ -8,60 +8,60 @@
...
+    std::vector<tablator::Data_Type> alt_datatypes;
+    bool format_ipac = (enum_format == Format::Enums::IPAC_TABLE);
+    bool format_votable = (enum_format == Format::Enums::VOTABLE);
+    bool format_fits = (enum_format == Format::Enums::FITS);
+    bool adjust_uint64 = (format_ipac || format_votable || format_fits);
 
-    // Skip the null bitfield flag
-    for (size_t col = 1; col <= columns.size(); ++col) {
-        if ((columns[col].type == tablator::Data_Type::UINT64_LE) && adjust_uint64) {
+    for (size_t col = 0; col <= columns.size(); ++col) {
```
Comment (smonkewitz) [inline src/Data_Type_Adjuster.cxx]: Wouldn't it be simpler to have the loop body retrieve the original type, adjust it to produce a new type, and only then append it to `alt_datatypes`, rather than appending first and then adjusting the `back()` of the vector?

File: src/Table/write_ipac_table/write_ipac_table.cxx [hunk 10]
```diff
@@ -8,92 +8,95 @@
...
+
 void tablator::Table::write_ipac_table(
-        std::ostream &os, const std::map<size_t, Data_Type> &datatypes_for_writing) const {
-    std::vector<size_t> ipac_column_widths = get_column_width();
+        std::ostream &os,
+        const std::map<size_t, Data_Type> &datatypes_for_writing) const {
+    std::vector<size_t> ipac_column_widths = get_column_widths();
     write_ipac_table_header(os);
+
+    std::vector<size_t> effective_num_arrays =
```
[the remark this answers is not in the document]
Comment (judith-ipac) [reply src/Table/write_ipac_table/write_ipac_table.cxx]: Already done---you must not have been seeing the very latest.

[the remark this answers is not in the document]
Comment (smonkewitz) [reply src/Table/write_ipac_table/write_ipac_table.cxx]: I am commenting in commit order, because the initial giant clang-format commit makes the overall diff basically unreadable. So there may be more of these sorts of comments - sorry!

**Comment by judith-ipac** [comment]
Oh dear.  I was hoping github would show you diffs  based on that commit.  Are you still working your way through commits?  Maybe not, if you approved the pull request. If you are, I’ll wait until you finish.

J

**Comment by smonkewitz** [comment]
Nope - I've read through all the commits.

**Comment by judith-ipac** [comment]
Good, thanks.

File: src/Table/write_ipac_table/write_ipac_table.cxx [hunk 11]
```diff
@@ -8,92 +8,95 @@
...
-                }
             }
         } else {
-            for (size_t element = 0; element < column.array_size; ++element) {
-                os << std::setw(*ipac_column) << unit->second << "|";
-                if (column.type == Data_Type::CHAR) {
-                    break;
-                }
+            for (size_t element = 0; element < effective_num_arrays[i]; ++element) {
+                os << std::setw(*width_ptr) << unit->second << "|";
```
[the remark this answers is not in the document]
Comment (judith-ipac) [reply src/Table/write_ipac_table/write_ipac_table.cxx]: That cleans up nicely, thanks.

File: src/Data_Type_Adjuster.cxx (hunk 9 above)
Comment (judith-ipac) [reply src/Data_Type_Adjuster.cxx]: I think I did that because it ensures that columns which keep orig_datatype are handled properly; otherwise they risk falling through cracks between the ifs.  Unless I'm missing something.

Comment (judith-ipac) [reply src/Data_Type_Adjuster.cxx]: I guess I could check the size of the alt_datatypes vector....

Comment (smonkewitz) [reply src/Data_Type_Adjuster.cxx]: Just to be clear, I mean something like:
```c++
for (size_t col = 0; col <= columns.size(); ++col) {
    auto orig_datatype = columns[col].type;
    auto adjusted_datatype = orig_datatype;
    if (salad) {
        ...
        adjusted_datatype = tablator::Data_Type::CHAR;
    } else {
        ...
    }
    alt_datatypes.emplace_back(adjusted_datatype); 
}
```
I don't see any cracks. (Famous last words?)

Comment (judith-ipac) [reply src/Data_Type_Adjuster.cxx]: Ah, yep, I was missing that.
